image/src/mobilenet/teachable-mobilenet.ts

```diff
@@ -133,25 +162,75 @@ export class TeachableMobileNet extends CustomMobileNet {
         return super.predict(image, flipped, maxPredictions);
     }
 
-
     /**
      * process the current examples provided to calculate labels and format
      * into proper tf.data.Dataset
      */
     public prepare() {
-        const xs: Float32Array[] = this.examples.map(ex => ex[1]);
-        if (!xs.length) {
-            throw new Error('Add some examples before training');
-        }
-        const ys: number[][] = [];
-        const numClasses = this.numClasses;
-        for ( const [label] of this.examples) {
-            ys.push(flatOneHot(label, numClasses));
+        for (let classes in this.examples){
+            if (classes.length == 0) {
+                throw new Error('Add some examples before training');
+            }
         }
 
-        this.dataset = convertToTfDataset(xs, ys);
+        const datasets = this.convertToTfDataset();
+        this.trainDataset = datasets.trainDataset;
+        this.validationDataset = datasets.validationDataset;
     }
 
+    /**
+     * Process the examples by first shuffling randomly per class, then adding
+     * one-hot labels, then splitting into training/validation datsets, and finally
+     * sorting one last time
+     */
+    private convertToTfDataset() {
```

Comment:
If I follow, this enforces that the ratio of classes in the training and validation sets is consistent with the ratio in the entire data set, is that right?  I read this as saying in a scenario with n = 20 and two classes with a split like `{a: 12, b: 8}`, this would lead to a training set of `{a: 10, b: 7}` and a validation set of `{a: 2, b: 1}`, where the ratio of examples is consistent.

In that example, which seems like it might be common for working with images, it sounds like the validation accuracy might be expected to swing pretty wildly when there are only three data points.  Because the training process isn't responsive to the validation accuracy (ie, it doesn't terminate early if validation accuracy is high, or say "model did not converge" if validation accuracy is low), there's sort of no harm in the split other than using less data for training.   But graphing validation accuracy would mostly show noise or large jumps that might not be super helpful pedagogically for visualizing 'progress' in learning across epochs.  

I think this essential constraint with datasets this small, where there is so much freedom for the classifier and everything kind of always overfits, so I'm just trying to check my understanding of what's happening here conceptually.  I do see that with grabbing samples from webcam frames the `n` might be higher than 20, and that validation graphs might be more meaningful, and either way it seems like a great improvement to support this splitting here!

As always, thanks for sharing such awesome work in the open 👍 and if these questions are distracting while work is in progress please let me know and I'll check my curiosity! 😄

Reply:
Hi Kevin - no worries! Love this discussion

So your intuition about what's happening with the splitting is correct. Read [this](https://developers.google.com/machine-learning/crash-course/training-and-test-sets/splitting-data) to learn more about that process. 

You're also right that there's no harm in splitting except for less data. It could certainly be a problem if there is TOO little data in the classes because then the model would have even less to learn from. Yes - I think there could perhaps be some confusion if someone is looking at the charts and doesn't know how to interpret them. That's why we're hoping to add some onboarding/tutorials around that. The loss function also tells an important story and can help you figure you what's going on. This is a [great article](https://machinelearningmastery.com/learning-curves-for-diagnosing-machine-learning-model-performance/) that explains how to interpret those a bit.

Does this answer your question? Comment?

Reply:
@irealva Awesome, yeah thanks for sharing the articles!

I think I'm on the same page about all those things, and wondering about what the training/validation accuracy graph might look like for an image classifier with say 20 training images.  But probably better to just try and see what it looks like :)

To that end, when something is merged in master here, does that mean it's deployed at the TM beta?  I see the new "under the hood bits" so I'm figuring yes.  Here's what I see for a dataset of 10 images of shiba inu dogs, and 10 images of maine coon cats:

<img width="393" alt="Screen Shot 2019-08-19 at 11 48 23 AM" src="https://user-images.githubusercontent.com/1056957/63279674-656c3d80-c277-11e9-89cc-2a6339a2705b.png">
<img width="396" alt="Screen Shot 2019-08-19 at 11 48 35 AM" src="https://user-images.githubusercontent.com/1056957/63279675-656c3d80-c277-11e9-91bc-93a67a9343d2.png">

My thought here is that the "training accuracy" graph in this scenario will never be super useful, because the number of validation examples will only be 2.  So the only possible validation accuracy values are `[0.0, 0.5 or 1.0]`.  And so using this graph to check for overfitting won't work the same way as it would for a larger dataset.

My overall assumption is that this is kind of essential tradeoff in these small datasets - the classifier has so much freedom to move around and still fit the constraint that it divides the two sets of examples (in other words, generalization will always be pretty bad).  So I'm just trying to think through how the kinds of articles like the ones you linked about overfitting apply to transfer learning with only ~20-100 examples in the training/validation set.  This is sort or related to what I was talking about over email with trying to visualize what exactly the TM model learns in that layer on top of MobileNet.

Reply:
I think with small datasets your intuition is right. But we've been testing with larger datasets like the popular flower one and there there can be differences that suggest the model is overfitting or not for example. 

Also, there might be other models we support that are not based on transfer learning.

And we have a staging environment! So not everything checked in here shows up immediately in the live app.

Reply:
Cool, thanks for sharing! 😄